src/site/pages/digital/tests/containers/ExprToCircuitPopup.test.tsx

```diff
@@ -0,0 +1,143 @@
+
+        // Check dropdowns
+        const inputOptions = screen.getByLabelText(/Input Component/).querySelectorAll("option");
+        const switchInputOption = [...inputOptions].find((input) => input.text === "Switch");
+        expect(switchInputOption?.selected).toBeTruthy();
+        const outputOptions = screen.getByLabelText(/Output Component/).querySelectorAll("option");
+        const ledOutputOption = [...outputOptions].find((input) => input.text === "LED");
+        expect(ledOutputOption?.selected).toBeTruthy();
+
+        // Text input is empty
+        const input = screen.getByPlaceholderText("!a | (B^third)") as HTMLInputElement;
+        expect(input.value).toBe("");
+    });
+
+    test("Cancel Button Cancels", async () => {
+        await user.type(screen.getByPlaceholderText("!a | (B^third)"), "a | b");
+
+        await user.click(screen.getByText("Cancel"));
+        expect(screen.getByText("Cancel")).not.toBeVisible();
+        expect(screen.getByText("Digital Expression To Circuit Generator")).not.toBeVisible();
+
+        // Reopen and requery in case reference changed
+        act(() => {store.dispatch(OpenHeaderPopup("expr_to_circuit"))});
+        expect((screen.getByPlaceholderText("!a | (B^third)") as HTMLInputElement).value).toBe("");
+    });
+
+    test("Generate Button", async () => {
+        // Enter the expression and generate
+        await user.type(screen.getByPlaceholderText("!a | (B^third)"), "a | b");
+        expect(screen.getByText("Generate")).toBeEnabled();
+        await user.click(screen.getByText("Generate"));
+        expect(screen.getByText("Digital Expression To Circuit Generator")).not.toBeVisible();
+
+        // Check that the components are placed and connected
+        const components = info.designer.getObjects();
+        expect(components).toHaveLength(4);
+        const inputA = components.find(component => component instanceof Switch
+                                                 && component.getName() === "a") as Switch;
+        const inputB = components.find(component => component instanceof Switch
+                                                 && component.getName() === "b") as Switch;
+        const orGate = components.find(component => component instanceof ORGate) as ORGate;
+        const led = components.find(component => component instanceof LED) as LED;
+        expect(inputA).toBeDefined();
+        expect(inputB).toBeDefined();
+        expect(orGate).toBeDefined();
+        expect(led).toBeDefined();
+        expect(led.isOn()).toBeFalsy();
+        inputA.click();
+        expect(led.isOn()).toBeTruthy();
+        inputA.click();
+        inputB.click();
+        expect(led.isOn()).toBeTruthy();
+
+        // Reopen and requery in case reference changed
+        act(() => {store.dispatch(OpenHeaderPopup("expr_to_circuit"))});
+        expect((screen.getByPlaceholderText("!a | (B^third)") as HTMLInputElement).value).toBe("");
```

Comment:
once again, is there a better way to query it? if not, better to maybe make a constant at the top since it's such a specific string?
also preferrable (if possible):
`screen.getByPlaceholderText<HTMLInputElement>("!a | (B^third)")`